contracts/rosca.sol

```diff
@@ -1,266 +1,258 @@
-            {
-                currentEpoch++;
-                currentRound = 1;
-                lowestBid = roundSum;
-
-            }
-
-
+          if(members[membersAddresses[(currentRound-1 + i) % membersAddresses.length]].paid < currentEpoch)
+          winnerAddress = membersAddresses[currentRound-1 + i];
         }
-
-    }
-    /**
-     * Registers another user-proxy for this ROSCA (msg.sender is the proxy).
-     */
-    function joinRequest() beforeStart
-    {
-        //only put the request in the pending list if they are not in the ROSCA already
-        if(!members[msg.sender].alive)
-            pendingJoinRequest[msg.sender] = true;
-        else
-            throw;
+      }
+      members[winnerAddress].pendingWithdrawl += lowestBid;
+      members[winnerAddress].paid++;
+      LogRoundFundsRelease(winnerAddress, lowestBid);
     }
 
-    function acceptJoinRequest(address requestor)
-        onlyForeman
-        beforeStart
+    if(currentRound < numRoundsInEpoch)  // reset variables related to bidding
     {
-        if(pendingJoinRequest[requestor])
-        {
-            members[requestor] = USER({paid: 0 , contributed: 0, alive: true, allowance: 0});
-            memberAddress.push(requestor);
-            numRounds++;
-            LogParticipantRegistered(requestor);
-            delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
-        }
-        else throw;
+      lowestBid = contributionSize;
+      winnerAddress = 0;
 
-
-    }
-
-    // v2: function leave();
-
-    /**
-     * Processes a periodic contribution from msg.sender (the user-proxy).
-     * Any excess funds will be withdrawable through withdraw().
-     */
-    function contribute() payable
+      currentRound++;
+    }else if(currentEpoch < numEpochs) //end of Epoch reset rounds
     {
-        //Product condsideration : are we gonna limit how much ether could be sent? is it total, or how much sent each time?
-        if(members[msg.sender].alive)
-        {
-            members[msg.sender].contributed = msg.value;
-            totalAmountInContract = msg.value;
-            //TODO(shine) : look into security about deposit and withdrawal
-            LogContributionMade(msg.sender, msg.value);
-        }
-        else
-           throw;
+      currentEpoch++;
+      currentRound = 1;
+      lowestBid = contributionSize;
     }
-
-    /**
-     * Registers a bid from msg.sender. If msg.sender has already won a round,
-     * this method will do nothing (or throw?).
-     */
-    function bid(uint distrubtionAmountInWei)
+  }
+  /**
+   * Anyone not already a member of ROSCA can request to join and they'll be put into
+   * pendingJoinRequest untill foreman accept request or ROSCA has started
+   *
+   */
+  function joinRequest() beforeStart
+  {
+    //only put the request in the pending list if they are not in the ROSCA already
+    if(!members[msg.sender].alive)
```

Comment:
consider removing the throw - if someone asks to join twice, it's not such a bad behavior, and by throwing, we penalize them by taking all available gas.
If you decide to leave the throw intact, I suggest changing to

if (!a) throw;
doSomething()

as it may be easier to read.